### PR TITLE
expression: rewrite builtin function: ANY_VALUE

### PR DESCRIPTION
**Pull request by jackysp**
For #4080 
PTAL @XuHuaiyu @zz-jason

### CODE REVIEW

File: expression/builtin_miscellaneous.go [hunk 1]
```diff
@@ -176,23 +183,130 @@ func (c *anyValueFunctionClass) getFunction(args []Expression, ctx context.Conte
...
+	case tpJSON:
+		sig = &builtinJSONAnyValueSig{baseJSONBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinRealAnyValueSig{baseRealBuiltinFunc{bf}}
+	case tpString:
+		sig = &builtinStringAnyValueSig{baseStringBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinTimeAnyValueSig{baseTimeBuiltinFunc{bf}}
+	default:
+		panic("unexpected evalTp")
```
Comment (zz-jason) [inline expression/builtin_miscellaneous.go]: panic message can be more specific, like: "unexpected evalTp of builtin function ANY_VALUE"

File: expression/builtin_miscellaneous.go [hunk 2]
```diff
@@ -176,23 +183,130 @@ func (c *anyValueFunctionClass) getFunction(args []Expression, ctx context.Conte
...
+	case tpJSON:
+		sig = &builtinJSONAnyValueSig{baseJSONBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinRealAnyValueSig{baseRealBuiltinFunc{bf}}
+	case tpString:
+		sig = &builtinStringAnyValueSig{baseStringBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinTimeAnyValueSig{baseTimeBuiltinFunc{bf}}
+	default:
+		panic("unexpected evalTp of builtin function ANY_VALUE")
```
Comment (shenli) [inline expression/builtin_miscellaneous.go]: Why not return error here?

Comment (jackysp) [inline expression/builtin_miscellaneous.go]: It should not happen, and there's no suitable error message for this error.

File: expression/builtin_miscellaneous.go [hunk 3]
```diff
@@ -176,23 +183,130 @@ func (c *anyValueFunctionClass) getFunction(args []Expression, ctx context.Conte
...
-// eval evals a builtinAnyValueSig.
+// evalDecimal evals a builtinDecimalAnyValueSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_any-value
-func (b *builtinAnyValueSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinDecimalAnyValueSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: why not `return val, isNull, errors.Trace(err)` directly?
so as the others

Comment (jackysp) [inline expression/builtin_miscellaneous.go]: Done. PTAL

Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: I mean that,
we do not need to check `if err != nil || isNull`,
we can `return val, isNull, errors.Trace(err)` directly.

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Pull request merged**